In [5]:
#hide
import numpy
import math
import pandas as pd
import scipy
from scipy.special import lambertw
from math import *
from numpy import *

#Problem 1-1. For each function f .n/ and time t in the following table, determine the largest
#size n of a problem that can be solved in time t , assuming that the algorithm to
#solve the problem takes f .n/ microseconds.

#You may assume that there are 30 days/month and 365 days /year. 



    
print("\nWe are creating the graph of different functions at different time intervals in microseconds\n"
        "This part has the math that I've used \n\n")
   #convert to microseconds - used google 
seconds = 10**6
minute = 6 * 10**7
hour = 3.6 * 10**9
day = 8.64 * 10**10
month = 2.628 * 10**12
year = 3.154 * 10**13
century = 3.154 * 10**15

listed = [10**6, 6 * 10**7, 3.6 * 10**9, 8.64 * 10**10,2.628 * 10**12, 3.154 * 10**13, 3.154 * 10**15]

 #turning it into a list so I can just loop through it 
time = []
time.append(seconds)
time.append(minute)
time.append(hour)
time.append(day)
time.append(month)
time.append(year)
time.append(century)
    
    
    #logn
logn = ["2^10^6","2^(6x10^7)","2^(3.6x10^9)","2^(8.64x10^10)","2^(2.628x10^12)","2^(3.154x10^13)","2^(3.154x10^15)"]
print("Log(n) = ",logn)
    
    #sqrt n
sqn = power(time,2)
sqrtn = list(map("{:.2f}".format,sqn))
print("Sq Root = ",sqrtn)
    
    
    #n
en = list(map("{:.2f}".format,time))
print("n = ",en)
    
#nlogn
#https://math.stackexchange.com/questions/406707/elegant-way-to-solve-n-log-2n-le-106
#https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.lambertw.html

#answers but without the 0.00j
nlogOn = ["62746","2801417","133378058","2755147513","71870856404","797633893349","68654697441062"]

def nlogen(num):
    return e**(lambertw(num*log(2)))

nlogn = []
for i in time:
    nlogen(i)
    nlogn.append(nlogen(i))
    #nlogn = nlogn.replace("+0.00j", "")
     
    
nlogn = list(map("{:.2f}".format,nlogn))
print("nLog(n) = ",nlogn)
   
    
    
#n^2[3*x for x in [111, 222, 333]]
two = sqrt(time)
ntwo = list(map("{:.2f}".format,two))
print("n^2 = ",ntwo)
    
    #n^3
three = power(time, 1/3)
nthree = list(map("{:.2f}".format,three))
print("n^3 = ",nthree)
    
    #2^n
twoFormal = ["19","25","31","35","41","44","51"]
two = log(time)/log(2)
twon = list(map("{:.2f}".format,two))          
print("2^n = ",twon)
    
    #n!
#listed is all int, time is string
#https://stackoverflow.com/questions/2652536/reverse-factorial
def fact(num):
    i = 0
    while True:
        if math.factorial(i) >= num:
            return i - 1
        else:
            i = i + 1
    
    
nfact = []
for i in listed:
    fact(i)
    nfact.append(fact(i))
    
print(nfact)   

print("n! = ",nfact)



#https://thispointer.com/python-pandas-how-to-get-column-and-row-names-in-dataframe/

num = [["2^10^6","2^(6x10^7)","2^(3.6x10^9)","2^(8.64x10^10)","2^(2.628x10^12)","2^(3.154x10^13)","2^(3.154x10^15)"],
        ['1.0E13', '3.6E15', '1.30E19', '7.46E21', '6.91E24', '9.95E26', '9.95R30'],
        ['1.0E7', '6.0E6', '3.6E9', '8.64E10', '2.63E12', '3.15E13', '3.15E15'],
        ['62746', '2801417', '133378058', '2755147513', '71870856404', '797633893349', '68654697441062'],
        ['1000.00', '7745.97', '60000.00', '293938.77', '1621110.73', '5616048.43', '56160484.33'],
        ['100.00', '391.49', '1532.62', '4420.84', '13799.87', '31595.16', '146651.74'],
        ['19.93', '25.84', '31.75', '36.33', '41.26', '44.84', '51.49'],
        ['9', '11', '12', '13', '15', '16', '17']
        ]
#os.system(jupyter nbconvert --to html Untitled.ipynb)
data = pd.DataFrame(num, columns = ["Second", "Minute", "Hour", "Day", "Month", "Year", "Century"], index = ["Log(n)", "SqRt(n)", "n", "nLog(n)", "n^2", "n^3", "2^n", "n!"])

#write to html
#https://www.geeksforgeeks.org/how-to-render-pandas-dataframe-as-html-table/
html = data.to_html()
text_file = open("TableProof.html", "w") 
text_file.write(html) 
text_file.close() 


We are creating the graph of different functions at different time intervals in microseconds
This part has the math that I've used 


Log(n) =  ['2^10^6', '2^(6x10^7)', '2^(3.6x10^9)', '2^(8.64x10^10)', '2^(2.628x10^12)', '2^(3.154x10^13)', '2^(3.154x10^15)']
Sq Root =  ['1000000000000.00', '3600000000000000.00', '12960000000000000000.00', '7464960000000000000000.00', '6906384000000000029360128.00', '994771599999999961609535488.00', '9947716000000000027312703668224.00']
n =  ['1000000.00', '60000000.00', '3600000000.00', '86400000000.00', '2628000000000.00', '31540000000000.00', '3154000000000000.00']
nLog(n) =  ['62746.13+0.00j', '2801417.88+0.00j', '133378058.86+0.00j', '2755147513.22+0.00j', '72830422138.38+0.00j', '797731502607.72+0.00j', '68619394449811.46+0.00j']
n^2 =  ['1000.00', '7745.97', '60000.00', '293938.77', '1621110.73', '5616048.43', '56160484.33']
n^3 =  ['100.00', '391.49', '1532.62', '4420.84', '13799.87', '31595.16', '146651.74']
2^n =  ['19.93', '25.84', '31.75',

In [2]:
import pandas as pd

#https://thispointer.com/python-pandas-how-to-get-column-and-row-names-in-dataframe/
print("Table Only\n")

num = [["2^10^6","2^(6x10^7)","2^(3.6x10^9)","2^(8.64x10^10)","2^(2.628x10^12)","2^(3.154x10^13)","2^(3.154x10^15)"],
        ['1.0E13', '3.6E15', '1.30E19', '7.46E21', '6.91E24', '9.95E26', '9.95R30'],
        ['1.0E7', '6.0E6', '3.6E9', '8.64E10', '2.63E12', '3.15E13', '3.15E15'],
        ['62746', '2801417', '133378058', '2755147513', '71870856404', '797633893349', '68654697441062'],
        ['1000.00', '7745.97', '60000.00', '293938.77', '1621110.73', '5616048.43', '56160484.33'],
        ['100.00', '391.49', '1532.62', '4420.84', '13799.87', '31595.16', '146651.74'],
        ['19.93', '25.84', '31.75', '36.33', '41.26', '44.84', '51.49'],
        ['9', '11', '12', '13', '15', '16', '17']
        ]
#os.system(jupyter nbconvert --to html Untitled.ipynb)
pd.DataFrame(num, columns = ["Second", "Minute", "Hour", "Day", "Month", "Year", "Century"], index = ["Log(n)", "SqRt(n)", "n", "nLog(n)", "n^2", "n^3", "2^n", "n!"])


Table Only



Second      Minute          Hour             Day            Month  \
Log(n)    2^10^6  2^(6x10^7)  2^(3.6x10^9)  2^(8.64x10^10)  2^(2.628x10^12)   
SqRt(n)   1.0E13      3.6E15       1.30E19         7.46E21          6.91E24   
n          1.0E7       6.0E6         3.6E9         8.64E10          2.63E12   
nLog(n)    62746     2801417     133378058      2755147513      71870856404   
n^2      1000.00     7745.97      60000.00       293938.77       1621110.73   
n^3       100.00      391.49       1532.62         4420.84         13799.87   
2^n        19.93       25.84         31.75           36.33            41.26   
n!             9          11            12              13               15   

                    Year          Century  
Log(n)   2^(3.154x10^13)  2^(3.154x10^15)  
SqRt(n)          9.95E26          9.95R30  
n                3.15E13          3.15E15  
nLog(n)     797633893349   68654697441062  
n^2           5616048.43      56160484.33  
n^3             31595.16        146651.74  
2^n                44.84            51.49  
n!                    16               17

In [7]:
#Kerry Gip
#CSCI 3412 003
#Due 2/21/21


#rand1000.txt
import time

def readData(numList):

    infileName = input("What is the name of the file you wish to enter?: ")  

    try:               
        infile = open(infileName, "r")
        for line in infile:
            for i in line.strip().split():
                numList.append(int(i))

       #Display error message if the input file is not found.     
    except FileNotFoundError:
        print("File not found")
    
    infile.close()
    
def insertionSort(numList):
    
    # https://www.geeksforgeeks.org/python-program-for-insertion-sort/
    # Traverse through 1 to len(numbers) 
    count = 0
    for i in range(1, len(numList)):
        
        #sets current position as sorted and remembers the first number of the list
        current = numList[i] 
        #position/index starts at 0
        position = i-1
        count = count +1
        
        #loops it all the way to the end of the list
        #if current is less than the numList at whatever position it is
        #position moves to the right
        while position >= 0 and current < numList[position] : 
                #swap
                numList[position + 1] = numList[position] 
                position = position -  1
                count = count +1
                #inserts unsorted element to correct position
        numList[position + 1] = current
        
        
    print("The number of insertion comparisons is: ", count)
    
def mergeSort(numList):
        #https://www.geeksforgeeks.org/merge-sort/
         # Traverse through 1 to len(numbers) 
    count = 0
    if len(numList) > 1:
 
         # Finding the middle of the list
        mid = len(numList)//2
 
        # Dividing the list into two halves
        leftList = numList[:mid]
        rightList = numList[mid:]
 
        #splitting the unsorted groups recursively until there is 1 element per group
        leftRecur = mergeSort(leftList)
        rightRecur = mergeSort(rightList)
         
        #also have to count the nmber of comparisons in the recursions   
        count += leftRecur + rightRecur
        
        indexLeft = 0
        indexRight = 0
        indexNumList = 0
        
 
        # while is just making sure the start position is less than length of list 
        while indexLeft < len(leftList) and indexRight < len(rightList):
            count = count +1
            
            #we are sorting the left list first before moving to the right list and moving towards the while loop
            if leftList[indexLeft] < rightList[indexRight]:
                #the original list index is now the index of the left list
                numList[indexNumList] = leftList[indexLeft]
                #index left moved over one and compares with the old numlist 
                indexLeft += 1
                
                #if on the right side of the list, then the original list index is now index of right list
            else:
                numList[indexNumList] = rightList[indexRight]
                indexRight += 1
                
            indexNumList += 1
            
        #handling leftovers on left side
        while indexLeft < len(leftList):
            numList[indexNumList] = leftList[indexLeft]
            indexLeft += 1
            indexNumList += 1
            count = count +1
            
        #handling leftovers on the right side
        while indexRight < len(rightList):
            numList[indexNumList] = rightList[indexRight]
            indexRight += 1
            indexNumList += 1
            count = count +1
    return count     
    #print("The number of merge comparisons is: ", count)


def timeEfficiency():
    #time.time or time.perf_counter()
    numList = []
    readData(numList)
    startIn = time.time()
    insertionSort(numList)
    endIn = time.time()
    
    
    #print results
    print("Insertion Starts at: ", startIn , "seconds")
    print("Insertion Ends at: ", endIn , "seconds")
    print("Insertion Time taken: ", endIn- startIn , "seconds")
    
    
    
    startMe = time.time()
    mergeSort(numList)
    endMe = time.time()
    
    
    #print results
    countMe = mergeSort(numList)
    print("\n\nThe number of merge comparisons is: ", countMe)
    print("Merge Starts at: ", startMe , "seconds")
    print("Merge Ends at: ", endMe , "seconds")
    print("Merge Time taken: ", endMe- startMe , "seconds")
    
timeEfficiency() 


What is the name of the file you wish to enter?: rand1000.txt
The number of insertion comparisons is:  247372
Insertion Starts at:  1613918450.764658 seconds
Insertion Ends at:  1613918450.8156688 seconds
Insertion Time taken:  0.051010847091674805 seconds


The number of merge comparisons is:  9976
Merge Starts at:  1613918450.8156688 seconds
Merge Ends at:  1613918450.8196697 seconds
Merge Time taken:  0.00400090217590332 seconds


This was ran on a windows 10 desktop, this completed before CSE Grid did and I ran it in the early afternoon on 2/19/21
My desktop build is 
CPU: AMD RYZEN 5 3600 
GPU: MSI RX5500XT
RAM: 16GB
Hopefully that will help people determine whether or not to run it locally or online


What is the name of the file you wish to enter?: rand1000.txt
The number of insertion comparisons is:  247372
Insertion Starts at:  1613771127.3292694
Insertion Ends at:  1613771127.3762798
Insertion Time taken:  0.04701042175292969


The number of merge comparisons is:  9976
Merge Starts at:  1613771127.3762798
Merge Ends at:  1613771127.3792808
Merge Time taken:  0.003000974655151367


What is the name of the file you wish to enter?: rand100000.txt
The number of insertion comparisons is:  2505319318
Insertion Starts at:  1613771460.440015
Insertion Ends at:  1613771949.9552004
Insertion Time taken:  489.51518535614014


The number of merge comparisons is:  1668928
Merge Starts at:  1613771949.9552004
Merge Ends at:  1613771950.4453332
Merge Time taken:  0.49013280868530273

What is the name of the file you wish to enter?: rand250000.txt
The number of insertion comparisons is:  15645054957
Insertion Starts at:  1613771986.19906
Insertion Ends at:  1613775251.3863704
Insertion Time taken:  3265.1873104572296


The number of merge comparisons is:  4487856
Merge Starts at:  1613775251.3863704
Merge Ends at:  1613775252.713671
Merge Time taken:  1.3273005485534668

What is the name of the file you wish to enter?: rand500000.txt
The number of insertion comparisons is:  62476143454
Insertion Starts at:  1613775659.2100985
Insertion Ends at:  1613790115.4757788
Insertion Time taken:  14456.26568031311


The number of merge comparisons is:  9475712
Merge Starts at:  1613790115.4787788
Merge Ends at:  1613790118.3904376
Merge Time taken:  2.911658763885498

What is the name of the file you wish to enter?: rand1000000.txt
The number of insertion comparisons is:  249926868108
Insertion Starts at:  1613838166.0206406
Insertion Ends at:  1613902299.3962376
Insertion Time taken:  64133.37559700012


The number of merge comparisons is:  19951424
Merge Starts at:  1613902299.3962376
Merge Ends at:  1613902305.680658
Merge Time taken:  6.284420490264893

PART 3 PROVING THAT MERGE SORT IS CORRECT



3.1  Write down a general description of the loop invariant technique in your own words as a proof technique of correctness.  (2 points)
A loop invariant is a statement that needs to be satisfied at the beginning of every loop
It also helps you understand why an algorithm gives you the correct answer

3.2 Identify the loop invariant of the loop in your merge() function (3 points)
My loop invariant is to make sure that the current(index) is less than the numberList of the index 
The outer loop: while indexLeft < len(leftList) and indexRight < len(rightList) 
is my loop invariant. It shows the start of the loop and that assumes that the numList[indexnumList] is currently empty or contains the smallest number 


3.3  Describe initialization step (0 points)
Initialization step is to make sure that the loop invariant is satisfied so it can continue looping

3.4  Describe maintenance  step (4 points)
If the loop invariant is true before the ith iteration, then the loop invariant will be true before the i + 1st iteration.
As the sort moves forward, it checks to make sure that the left number is smaller than the right number, then moves onto the next position and compares that one. 
It goes numList[indexNumList] = leftList[indexLeft] where it re-establishes that the numList[indexnumlist] now contains the smallest numbers in the list


3.5  Describe Termination step (1 point)
When the loop terminates, the loop invariant shows that it is true and that helps show that the algorithm is correct
Basically when any of the while loop conditions can no longer be fulfilled, 
while indexLeft < len(leftList) and indexRight < len(rightList) 
is when this can no longer be satisfied and the largest numbers are now added to the list




Extra Credit work!
*Open GipKerry Extra Credit first

From what I noticed, it looks like mergesort is the way to go. It is consistently faster and has consistent times regardless of size than insertion sort as the number of elements increase. Insertion sort grows exponentially in time and comparisons as it has to constantly compare itself against each element in the sorted list. I did not get it to completely run on CSE Grid. Insertion sort took 17.8 hours to sort the 1 million elements on my desktop, 4 hours for 500k, a little less than an hour for 250k and a few minutes for 100k. Merge sort continuously just took seconds. Even at 1 million elements, merge sort took 6 seconds. Consistent running times seems like it is the way to go. Even at 1000 elements, insertion sort was 1/10th slower than merge sort, despite it supposed to be faster at lower elements. 

This was ran on a windows 10 desktop, this completed before CSE Grid did and I ran it in the early afternoon on 2/19/21
My desktop build is 
CPU: AMD RYZEN 5 3600 
GPU: MSI RX5500XT
RAM: 16GB
Hopefully that will help people determine whether or not to run it locally or online